Deploy self-hosted (#418)
sunodo/sunodo: apps/docs/guide/deploying/self-hosted.md

```shell
$ sunodo deploy --hosting self-hosted
? Cartesi machine templateHash 0xc87999b8a93609268b10de25f2e49d35f80fad92813310edc585ed644a9805d3
? Application node Docker image sha256:5c355a9bddc92aa08987f395a257a0b32a51552c969eb161386e46f9380ea2ac
```

The image produced, in this example `sha256:5c355a...0ea2ac`, can be tagged and pushed to a Docker registry of choice, like Docker Hub, or any other private or public registry.

The next step is to deploy the smart contract for the application. The deploy command will redirect the developer to a web application, where he can use a wallet to deploy the required smart contracts.

![deploy self-hosted](./self-hosted-web.png)

## Wallet

The developer will need a wallet that will be used by the Cartesi rollups node to submit transactions to the base layer. This wallet should ideally be used exclusively for that purpose.

A new wallet can be created using a tool like [cast](https://book.getfoundry.sh/reference/cast/cast-wallet-new-mnemonic). A wallet managed by [AWS KMS](https://aws.amazon.com/blogs/database/part1-use-aws-kms-to-securely-manage-ethereum-accounts/) is also supported for an increased level of security.

::: warning

> The deploy web application need the wallet PUBLIC ADDRESS only!

Reviewer: ```suggestion
The deploy web application needs the wallet PUBLIC ADDRESS only!
```
This sentence is a bit weird to read. I guess "deploy web application" is the name of the web app but sounds a bit strange. Maybe "deployment web application" or "web app for deployment"

DON'T provide your mnemonic passphrase or private key.
:::

After the successful execution of the deployment transaction the developer is presented with information for the execution of the rollups node:

![deploy self-hosted config](./self-hosted-deployed-env.png)

The configuration for the node is presented in a `.env` file format. It includes the addresses of the deployed smart contracts, as well as information of the base layer chain.

> Three pieces are information are still missing though: the database connection string, the web3 provider URLs, and the wallet private key. The developer can download the configuration as an `.env` file and fill in the missing information.

Reviewer: ```suggestion
Three pieces of information are still missing though: the database connection string, the web3 provider URLs, and the wallet private key. The developer can download the configuration as an `.env` file and fill in the missing information.
```

:::

## PostgreSQL database

The rollups node requires a PostgreSQL database to store information about the application state, which is served via the GraphQL API.

The developer can use any PostgreSQL database, either managed by a cloud provider, or run on his own infrastructure. The only configuration that must be provided is the connection string, which includes the database URL, username, password, and database name. If the developer decides to host his own database he must take care of security and maintenance, like regular backups and updates.

The connection string must be configured as an environment variable called `CARTESI_POSTGRES_ENDPOINT`.

::: warning
The database connection string, which includes the password, is a sensitive information. The developer should take care to keep it secure, like using a secret mechanism provided by the cloud provider of choice.
:::

The developer doesn't need to create the database schema, as the rollups node will do it automatically when it starts.

## Web3 provider

> The rollups node need a connection to base layer through a standard JSON-RPC API endpoint. It requires both a HTTP endpoint as well as a WebSocket endpoint.

Reviewer: ```suggestion
The rollups node needs a connection to the base layer through a standard JSON-RPC API endpoint. It requires both an HTTP endpoint as well as a WebSocket endpoint.
```

The Cartesi rollups node is distributed as a Docker image, so the server will need to have Docker installed. Any popular cloud provider, like AWS, GCP, Azure, Digital Ocean, or Linode, is capable of running docker containers, either manually by spawing a server and running docker, or by using a managed container infrastructure like Kubernetes.

The developer can also use a service like [Fly.io](https://fly.io) to deploy the application node.

We will describe below two methods of running a rollups node, the first using Docker, which can be used as the basic information to use any cloud provider, and the second using [Fly.io](https://fly.io), a managed service for running Docker containers.

### Using Docker in any cloud provider

Assuming the developer completed the above requirements, and has a complete `.env` file, he can run the rollups node using the following command:

```shell
$ docker run --env-file <env-file> -p 10000:10000 <image-id>
(...)
INFO http: listening at [::]:10000
(...)
INFO rollups-node: all services are ready
(...)
```

> In the examples above the `<env-file>` file is `0xc87999b8a93609268b10de25f2e49d35f80fad92813310edc585ed644a9805d3.env` and the `<image-id>` is `sha256:5c355a9bddc92aa08987f395a257a0b32a51552c969eb161386e46f9380ea2ac`.

Reviewer 2: Maybe a disclaimer the developer need top push the image to a container registry accessible by the cloud provider where the command will be executed.

The developer is free to use any managed container solution, like Kubernetes. The rollups node is just a single container, with a single exposed port, that must be accessible from the internet.

### Using fly.io

> [Fly.io](https://fly.io) is platform as a service that allows developers to easily deploy applications packaged as Docker containers. The rollups node can be deployed to Fly.io. Choose an `<app-name>` and follow the steps below:

Reviewer: ```suggestion
[Fly.io](https://fly.io) is a platform as a service that allows developers to easily deploy applications packaged as Docker containers. To deploy the rollups node to Fly.io, choose an `<app-name>` and follow the steps below:
```
Weird sentence, changed it a bit

1. [Install the flyctl CLI](https://fly.io/docs/hands-on/install-flyctl/)
2. [Create an account and login](https://fly.io/docs/hands-on/sign-up-sign-in/)
3. Create an application:

    ```shell

>     $ fly app create <app-name>
>     New app created: <app-name>

Reviewer 2: If the idea is that the use copy and paste the commands, we could define a `$APP_NAME` here and reuse.

```suggestion
    $ export APP_NAME=<app-name>
    $ fly app create $APP_NAME
    New app created: <app-name>
```

    ```

4. Create a postgres database application:

    ```shell
    fly postgres create --initial-cluster-size 1 --name <app-name>-database --vm-size shared-cpu-1x --volume-size 1
    ```

5. Attach database to the node application:

>     The previous step should have printed the connection string for the database. The developer can use it to configure the `CARTESI_POSTGRES_ENDPOINT` environment variable.

Reviewer: Why is this sentence here? I understand that the code to attach has nothing to do with this connection string, right? Maybe we should place this sentence at the end of step 4 (e.g., "take note of the connection string to use it later on")

    ```shell
    fly postgres attach <app-name>-database -a <app-name>
    ```

6. Download `fly.toml` file from deployment:

    ![deploy self-hosted config](./self-hosted-deployed-fly.png)

7. Create secrets for sensitive configuration, replacing with the actual values:

>     ```shell
>     fly secrets set -a <app-name> CARTESI_BLOCKCHAIN_HTTP_ENDPOINT=<web3-provider-http-endpoint>
>     fly secrets set -a <app-name> CARTESI_BLOCKCHAIN_WS_ENDPOINT=<web3-provider-ws-endpoint>
>     fly secrets set -a <app-name> CARTESI_AUTH_MNEMONIC=<mnemonic>
>     fly secrets set -a <app-name> CARTESI_POSTGRES_ENDPOINT=<connection_string>
>     ```

Reviewer 2: `fly secrets` can receive a list if you wanna make it less verbose

```suggestion
    fly secrets set -a <app-name> \
        CARTESI_BLOCKCHAIN_HTTP_ENDPOINT=<web3-provider-http-endpoint> \
        CARTESI_BLOCKCHAIN_WS_ENDPOINT=<web3-provider-ws-endpoint> \
        CARTESI_AUTH_MNEMONIC=<mnemonic> \
        CARTESI_POSTGRES_ENDPOINT=<connection_string>
```

Author: I any case that is not much "copy and pastable", because of the need to replace the values between `<>`
